server/gendb/interface.go.tmpl

```diff
@@ -108,10 +108,27 @@ type Scan{{ $modelName }}sInput struct {
 {{- $indexName := indexName $xdbConfig.DynamoDB.KeySchema }}
 {{- $hashKey := index $xdbConfig.DynamoDB.KeySchema 0 }}
 {{- $pkModelAttributeNames := modelAttributeNamesForIndex $xdbConfig $xdbConfig.DynamoDB.KeySchema }}
+{{- $nonIndexAttributeNames := nonIndexModelAttributeNames $xdbConfig $pkModelAttributeNames}}
+{{- $lenNonIndexAttributes := len $nonIndexAttributeNames }}
 {{- if indexHasRangeKey $xdbConfig.DynamoDB.KeySchema }}
 {{- $rangeKey := index $xdbConfig.DynamoDB.KeySchema 1 }}
 {{- $rangeKeyAttributes := modelAttributeNamesForKeyType $xdbConfig $xdbConfig.DynamoDB.KeySchema "RANGE" }}
 {{- $lenRangeKeyAttributes := len $rangeKeyAttributes }}
+{{- if gt $lenNonIndexAttributes 0 }}
+    // {{ $modelName }}By{{ $indexName }}FilterableAttribute represents the fields we can apply filters to for queries on this index
+    type {{ $modelName }}By{{ $indexName }}FilterableAttribute string
+    {{- range $attributeName := $nonIndexAttributeNames }}
+      const {{ $modelName }}{{ pascalize $attributeName }} {{ $modelName }}By{{ $indexName }}FilterableAttribute = "{{ $attributeName }}"
+    {{- end }}
+
+    // {{ $modelName }}By{{ $indexName }}Filter represents a filter on a particular field to be included in the query
+    type {{ $modelName }}By{{ $indexName }}FilterValues struct {
+      // AttributeName is the attibute we are attempting to apply the filter to
+      AttributeName {{ $modelName }}By{{ $indexName }}FilterableAttribute
+      // AttributeValues is an optional parameter to be used when we want to compare the attibute to a single value or multiple values
+      AttributeValues []interface{}
```

Comment:
What would it mean to not have `AttributeValues` set? Wondering since the comment says it is optional

Reply:
Most likely it would mean that the filter is checking for the existence of the field in the record.  In that case we wouldn't need any reference to a value just the field name itself